node: devicemgr: docs: Kubelet Device Manager graduation to GA

swatisehgal opened the pull request:
Docs update for Kubelet Device Manager graduation to GA.
Issue: https://github.com/kubernetes/enhancements/issues/3573

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

sftim commented on content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:
```diff
@@ -164,16 +164,25 @@ Pod onto Nodes, to restart the daemon Pod after failure, and to help automate up
+guaranteed to be non-breaking.
 
-* Watch for changes in future releases.
+NOTE: Graduation of Device Manager to a stable feature in v1.26 does not imply
+graduation of device plugin API to a stable API.
```
This should be something like:
```markdown
{{< caution >}}
Although the Device Manager component of Kubernetes is a generally available feature,
the _device plugin API_ is not stable. For information on the device plugin API and
version compatibility, read [Device Plugin API versions](/docs/reference/node/device-plugin-api-versions/).
{{< caution >}}
```
:warning: To make that text make sense, you (or another contributor) must write a PR that adds
the page https://k8s.io/docs/reference/node/device-plugin-api-versions/

swatisehgal replied:
Ack, will fix. Thanks

sftim commented on content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:
```diff
@@ -164,16 +164,25 @@ Pod onto Nodes, to restart the daemon Pod after failure, and to help automate up
 
-Kubernetes device plugin support is in beta. The API may change before stabilization,
-in incompatible ways. As a project, Kubernetes recommends that device plugin developers:
+Prior to v1.10, the versioning scheme required the Device Plugin's API version to
+match exactly the Kubelet's version. On graduation of this feature to Beta in v1.12
+this is no longer a hard requirement. The API is versioned and has been stable since
+Beta graduation of this feature. Because of this, Kubelet upgrades should be seamless
+but there still may be changes in the API before stabilization making upgrades not
+guaranteed to be non-breaking.
```
We can confidently assume that the reader is running Kubernetes newer than v1.13 - see https://kubernetes.io/docs/home/supported-doc-versions/

sftim reviewed (commented):
Thanks for this PR @swatisehgal

I think we should move some of the details of the device plugin API into the reference section. For graduation to stable, we want docs that follow the website style guide and its structure (which is inspired by [Diátaxis](https://diataxis.fr/)).

Try to focus the content of the existing concept page on explaining what device plugins _are_.

See inline feedback for a specific example of content I would move. There's likely to be more though.

Also:
- does https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/ need an update?
- does https://kubernetes.io/docs/tasks/administer-cluster/memory-manager/#device-plugin-resource-api need to change its hyperlink to a more specific link, or perhaps to a new reference page?
- https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#handling-kubelet-restarts mentions a “current implementation”; is this now a defined and stable behavior that authors can rely on?

swatisehgal commented:
Thanks @sftim for your comments. I have tried to address your review comments, please take a look and let me know if you have addtional feedback.
> Thanks for this PR @swatisehgal
> 
> I think we should move some of the details of the device plugin API into the reference section. For graduation to stable, we want docs that follow the website style guide and its structure (which is inspired by [Diátaxis](https://diataxis.fr/)).
> 
> Try to focus the content of the existing concept page on explaining what device plugins _are_.
> 
> See inline feedback for a specific example of content I would move. There's likely to be more though.
> Also:
> 
> * does https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/ need an update?

Thanks, updated the k8s_version and state. 
> * does https://kubernetes.io/docs/tasks/administer-cluster/memory-manager/#device-plugin-resource-api need to change its hyperlink to a more specific link, or perhaps to a new reference page?

Updated the link to https://kubernetes.io//docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#grpc-endpoint-list and updated the text.
> * https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#handling-kubelet-restarts mentions a “current implementation”; is this now a defined and stable behavior that authors can rely on?

Yes, this matches with current behavior: https://github.com/kubernetes/kubernetes/blob/release-1.25/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go#L93

sftim commented on content/en/docs/tasks/manage-gpus/scheduling-gpus.md:
```diff
@@ -8,9 +8,9 @@ description: Configure and schedule GPUs for use as a resource by nodes in a clu
+{{< feature-state state="stable" for_k8s_version="v1.26" >}}
 
-Kubernetes includes **experimental** support for managing GPUs
+Kubernetes includes **stable** support for managing GPUs
 (graphical processing units) across several nodes.
```
Optional extra tweak:
```suggestion
Kubernetes supports managing GPUs (graphical processing units) across
different nodes in your cluster, using
{{< glossary_tooltip text="device plugins" term_id="device-plugin" >}}.
```
(and then remove the later use of that glossary tooltip shortcode).

sftim commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,29 @@
+  change significantly in a single release.
+* `-` The Kubernetes cluster has features the device plugin API can't use,
+  either due to the server has additional new API, or that device plugin API has
+  removed old API. However, everything they share in common (i.e., most APIs)
```
```suggestion
removed old API. However, everything they share in common (most APIs)
```

sftim commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,29 @@
+* `+` device plugin API has features or API objects that may not be present in the
+  Kubernetes cluster, either due to that device plugin API has additional new API, or
+  that the server has removed old API. However, everything they have in
+  common (i.e., most APIs) will work. Please note that alpha APIs may vanish or
```
```suggestion
  common (most other APIs) will work. Note that alpha APIs may vanish or
```

sftim commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,29 @@
+  Kubernetes cluster, either due to that device plugin API has additional new API, or
+  that the server has removed old API. However, everything they have in
+  common (i.e., most APIs) will work. Please note that alpha APIs may vanish or
+  change significantly in a single release.
```
```suggestion
  change significantly between one minor release and the next.
```
?

swatisehgal replied:
Sure, updated.

sftim commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,29 @@
+
+#### Compatibility matrix
+
+|               |  Kubernetes 1.21 | Kubernetes 1.22 | Kubernetes 1.23 | Kubernetes 1.24 | Kubernetes 1.25 | Kubernetes 1.26 |
```
I recommend flipping this table around: columns for alpha, beta, etc and rows for Kubernetes minor versions.

That will make it easier to maintain.

swatisehgal replied:
Sure, updated.

sftim commented on content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:
```diff
@@ -164,16 +164,28 @@ Pod onto Nodes, to restart the daemon Pod after failure, and to help automate up
-a Kubernetes release with a newer device plugin API version, upgrade your device plugins
-to support both versions before upgrading these nodes. Taking that approach will
-ensure the continuous functioning of the device allocations during the upgrade.
+To run device plugins on nodes that need to be upgraded to a Kubernetes release with
+a newer device plugin API version, upgrade your device plugins to support both versions
+before upgrading these nodes. Taking that approach will ensure the continuous functioning
+of the device allocations during the upgrade.
```
We should also add this advice (or a link to this advice) into https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/

swatisehgal replied:
Added a note for device plugins and links for details information.

sftim commented on content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:
```diff
@@ -164,16 +164,28 @@ Pod onto Nodes, to restart the daemon Pod after failure, and to help automate up
+Previously, the versioning scheme required the Device Plugin's API version to match
+exactly the Kubelet's version. Since the graduation of this feature to Beta in v1.12
+this is no longer a hard requirement. The API is versioned and has been stable since
+Beta graduation of this feature. Because of this, Kubelet upgrades should be seamless
```
(nit)
```suggestion
Beta graduation of this feature. Because of this, kubelet upgrades should be seamless
```

sftim reviewed (commented):
LGTM (with suggested extra improvements) for docs.

This is now also ready for a technical review by SIG Node folks.

swatisehgal commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,29 @@
+weight: 10
+---
+
+Please note that the device plugin API is present here: https://github.com/kubernetes/kubelet/tree/master/pkg/apis/deviceplugin.
```
Fixed, Thanks.

sftim commented on content/en/docs/tasks/administer-cluster/cluster-upgrade.md:
```diff
@@ -91,3 +91,12 @@ kubectl convert -f pod.yaml --output-version v1
+both version before node is upgraded for continuous functioning of the device allocations
+during the upgrade.
+
+Please refer to [API compatiblity](docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md/#api-compatibility) and [Kubelet Device Manager API Versions](docs/reference/node/device-plugin-api-versions.md) for more details.
```
(nit)
```suggestion
Refer to [API compatiblity](docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md/#api-compatibility) and [Kubelet Device Manager API Versions](docs/reference/node/device-plugin-api-versions.md) for more details.
```

sftim reviewed (commented):
I just noticed that this PR doesn't update https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ - it should.

swatisehgal commented:
> I just noticed that this PR doesn't update https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ - it should.

Thanks for that, updated.

sftim commented:
LGTM

Ready for tech review (which should be straightforward, I think).

swatisehgal commented:
/cc @fromanirh @klueska 
for tech review

fromanirh commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -251,6 +249,9 @@ For a reference to old feature gates that are removed, please refer to
 | `DefaultPodTopologySpread` | `true` | Beta | 1.20 | 1.23 |
 | `DefaultPodTopologySpread` | `true` | GA | 1.24 | - |
+| `DevicePlugins` | `false` | Alpha | 1.8 | 1.9 |
+| `DevicePlugins` | `true` | Beta | 1.10 | |
```
asking, not sure myself: should this read like
```
`DevicePlugins` | `true` | Beta | 1.10 | 1.25 |
```
?

fromanirh replied:
Additionally, since we're graduating to GA, do we need to keep this content in this file or should we move this into the `feature-gates-removed.md` file? Again asking, I'm not sure myself about the process.

swatisehgal replied:
Going to go ahead and add the following (as I think it should have been added initially itself):
```
DevicePlugins` | `true` | Beta | 1.10 | 1.25 |
```
Will wait for input on whether this needs to go to into the `feature-gates-removed.md` file or not.

swatisehgal replied:
Found this text https://github.com/kubernetes/website/blame/main/content/en/docs/reference/command-line-tools-reference/feature-gates-removed.md#L9-L18 that indicates that we shouldn't have feature gates corresponding to GA features in `feature-gates-removed.md`.

fromanirh reviewed (commented):
/lgtm
from sig-node, for technical content. Couple question inside about process.
I recommend to wait for the review of @klueska (if possible) because he's a maintainer of the related code since long time and can be aware of details I'm missing.

sftim reviewed (commented):
/hold

This PR should target the _dev-1.26_ branch, right? OK to unhold once it does.

klueska commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,34 @@
+  version.
+* `+` device plugin API has features or API objects that may not be present in the
+  Kubernetes cluster, either due to that device plugin API has additional new API, or
+  that the server has removed old API. However, everything they have in
```
```suggestion
  that the server has removed an old API call. However, everything they have in
```

swatisehgal replied:
Updated, Thanks.

klueska commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,34 @@
+  common (most other APIs) will work. Note that alpha APIs may vanish or
+  change significantly between one minor release and the next.
+* `-` The Kubernetes cluster has features the device plugin API can't use,
+  either due to the server has additional new API, or that device plugin API has
```
```suggestion
  either because server has added additional API calls, or that device plugin API has
```

swatisehgal replied:
Updated, Thanks.

klueska commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,34 @@
+  change significantly between one minor release and the next.
+* `-` The Kubernetes cluster has features the device plugin API can't use,
+  either due to the server has additional new API, or that device plugin API has
+  removed old API. However, everything they share in common (most APIs) will work.
```
```suggestion
  removed an old API call. However, everything they share in common (most APIs) will work.
```

swatisehgal replied:
Updated, Thanks.

klueska commented on content/en/docs/tasks/administer-cluster/cluster-upgrade.md:
```diff
@@ -91,3 +91,12 @@ kubectl convert -f pod.yaml --output-version v1
+
+### Device Plugins
+
+If your cluster is running device plugins and the node needs to been upgraded to a Kubernetes
```
```suggestion
If your cluster is running device plugins and the node needs to be upgraded to a Kubernetes
```

swatisehgal replied:
Updated, Thanks.

klueska commented on content/en/docs/tasks/administer-cluster/cluster-upgrade.md:
```diff
@@ -91,3 +91,12 @@ kubectl convert -f pod.yaml --output-version v1
+
+If your cluster is running device plugins and the node needs to been upgraded to a Kubernetes
+release with a newer device plugin API version, device plugins must be upgraded to support
+both version before node is upgraded for continuous functioning of the device allocations
```
```suggestion
both version before the node is upgraded in order to guarantee that device allocations continue to complete successfully 
```

swatisehgal replied:
Updated, Thanks.

klueska commented on content/en/docs/tasks/administer-cluster/memory-manager.md:
```diff
@@ -459,8 +459,9 @@ by using `--reserved-memory` flag.
 
-By employing the [API](/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/),
-the information about reserved memory for each container can be retrieved, which is contained
+Kubelet provides a `PodResourceLister` gRPC service to enable discovery of resources and associated metadata.
```
```suggestion
The kubelet provides a `PodResourceLister` gRPC service to enable discovery of resources and associated metadata.
```

swatisehgal replied:
Updated, Thanks.

klueska commented on content/en/docs/tasks/manage-gpus/scheduling-gpus.md:
```diff
@@ -23,8 +24,7 @@ and the current limitations.
 ## Using device plugins
 
-Kubernetes implements {{< glossary_tooltip text="Device Plugins" term_id="device-plugin" >}}
-to let Pods access specialized hardware features such as GPUs.
+Kubernetes implements device plugins to let Pods access specialized hardware features such as GPUs.
```
Is there a reason for this change?

swatisehgal replied:
In the overview section above we provided a link to the device plugin page when it is referred for the first time so removed that from here.

swatisehgal commented on content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:
```diff
@@ -164,16 +164,28 @@ Pod onto Nodes, to restart the daemon Pod after failure, and to help automate up
+this is no longer a hard requirement. The API is versioned and has been stable since
+Beta graduation of this feature. Because of this, kubelet upgrades should be seamless
+but there still may be changes in the API before stabilization making upgrades not
+guaranteed to be non-breaking.
```
Since we are graduating Device Manager component to GA and not the API, I captured this text here.

swatisehgal commented on content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:
```diff
@@ -164,16 +164,28 @@ Pod onto Nodes, to restart the daemon Pod after failure, and to help automate up
+guaranteed to be non-breaking.
 
-* Watch for changes in future releases.
+{{< caution >}}
+Although the Device Manager component of Kubernetes is a generally available feature,
+the _device plugin API_ is not stable. For information on the device plugin API and
+version compatibility, read [Device Plugin API versions](/docs/reference/node/device-plugin-api-versions/).
+{{< caution >}}
```
The reason I thought we need to split the graduation of the API and the component is because I noticed that there are proposals in the community to further enhance the device plugin API e.g. [Add of `Deallocate` and `PostStopContainer` to Device Manager API](https://github.com/kubernetes/enhancements/pull/3080). I am not 100% sure if there is still room for enhancements like this in the future if we were to mark the API stable and wanted to to avoid preventing future enhancements.

Having said that, Device Plugin API has been reasonably stable for a while now and with newer proposal such as dynamic resource allocation tackling a broader range of device related use case, I understand if we want to graduate the API along with the component. I am happy to make updates taking this input into consideration.

fromanirh replied:
I think that @swatisehgal 's reasoning is sound especially considering we're discussing `Deallocate` for a couple cycles. So splitting this work (GA'ing device manager proper and GA'ing the API) makes sense. 
I'm not sure about the process here. If the Device Manager API can be extended without too much hassle once stable (e.g.: would that require a separate graduation history?) it makes sense to GA the API alongside the DeviceManager proper in this cycle. Otherwise it's perhaps more practical to postpone the GA graduation of the API.
If we pursue the latter, it could make sense to put a deadline here to avoid perma-beta of the API as well. I'd say NEXT+2 top (so 1.28) but I'm brainstorming now.

swatisehgal replied:
We agreed that it is okay to postpone the graduation of device plugin API to a later stage. Slack thread with the discussion on this can be found [here](https://kubernetes.slack.com/archives/C0BP8PW9G/p1667481109075849).

swatisehgal commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,34 @@
+* `✓` Exactly the same features / API objects in both device plugin API and
+   the Kubernetes
+  version.
+* `+` device plugin API has features or API objects that may not be present in the
```
Updated, Thanks.

tengqm commented on content/en/docs/reference/node/device-plugin-api-versions.md:
```diff
@@ -0,0 +1,34 @@
+Key:
+
+* `✓` Exactly the same features / API objects in both device plugin API and
+   the Kubernetes
+  version.
```
```suggestion
   the Kubernetes version.
```

swatisehgal replied:
Thanks, fixed.

tengqm commented:
Target branch changed.
/hold cancel

fromanirh commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -88,8 +88,6 @@ different Kubernetes components.
 | `DelegateFSGroupToCSIDriver` | `false` | Alpha | 1.22 | 1.22 |
 | `DelegateFSGroupToCSIDriver` | `true` | Beta | 1.23 | |
-| `DevicePlugins` | `false` | Alpha | 1.8 | 1.9 |
-| `DevicePlugins` | `true` | Beta | 1.10 | |
```
It seems we have a new document in `command-line-tools-reference/` tracking the removed feature gate. This is likely the reason why the bot is complaining about merge conflict

swatisehgal replied:
Thanks, rebased.

fromanirh reviewed (commented):
/lgtm
from sig-node perspective. All the comments have been addressed

swatisehgal commented:
@sftim @tengqm Could you please take a look at this PR and see if all looks good to you.

nate-double-u commented:
/cc @Rishit-dagli @krol3
